File: 1002-find-common-characters.java
class Solution {
    public List<String> commonChars(String[] words) {
        Map<Character,Integer> freq1 = new HashMap<>();
        for(Character ch:words[0].toCharArray()){
            freq1.put(ch,freq1.getOrDefault(ch,0)+1);
        }
        
        for(String word : words){
            Map<Character,Integer> freq2 = new HashMap<>();
            for(Character ch:word.toCharArray()){
                freq2.put(ch,freq2.getOrDefault(ch,0)+1);
            }
            for(Character ch:freq1.keySet()){
                if(freq2.containsKey(ch)){
                    freq1.put(ch,Math.min(freq1.get(ch),freq2.get(ch)));
                }
                else freq1.put(ch,0);
            }
        }

        List<String> ans = new ArrayList<>();
        for(Character ch:freq1.keySet()){
            for(int i=0;i<freq1.get(ch);i++){
                ans.add(String.valueOf(ch));
            }
        }
        return ans;
    }
}

File: 1002-find-common-characters.py
class Solution:
    def commonChars(self, words: List[str]) -> List[str]:
        freq1 = {}
        for char in words[0]:
            freq1[char] = freq1.get(char,0) + 1
        for i in range(1,len(words)):
            freq2 = {}
            for char in words[i]:
                freq2[char] = freq2.get(char,0) + 1
            for char in freq1:
                if char in freq2:
                    freq1[char] = min(freq1[char], freq2[char])
                else:
                    freq1[char] = 0

        ans = []
        for x,num in freq1.items():
            for i in range(num):
                ans.append(x)

        return ans